airbyte-protocol/protocol-models/src/main/java/io/airbyte/protocol/models/JsonSchemaType.java

```diff
@@ -109,12 +113,25 @@ public static class Builder {
     private Builder(final JsonSchemaPrimitive type) {
       typeMapBuilder = ImmutableMap.builder();
       if (JsonSchemaPrimitiveUtil.isV0Schema(type)) {
-        typeMapBuilder.put(TYPE, type.name().toLowerCase());
+        if (type.equals(JsonSchemaPrimitive.JSONB)) {
+          buildJsonbSchema();
+        } else {
+          typeMapBuilder.put(TYPE, type.name().toLowerCase());
+        }
       } else {
         typeMapBuilder.put(REF, PRIMITIVE_TO_REFERENCE_BIMAP.get(type));
       }
     }
 
+    private void buildJsonbSchema() {
```

Comment:
random thought - can you try with this schema?
```json
{
  "type": ["object", "array", "string", "number", "boolean"],
  "airbyte_type": "jsonb"
}
```

I think that avoids the misleading `type: object`, and maybe works better for normalization

(sorry, I didn't think of this until now)

Reply:
> random thought - can you try with this schema?
> 
> ```json
> {
>   "type": ["object", "array", "string", "number", "boolean"],
>   "airbyte_type": "jsonb"
> }
> ```
> 
> I think that avoids the misleading `type: object`, and maybe works better for normalization
> 
> (sorry, I didn't think of this until now)

@edgao Yes, I tried this schema but the results are worse then
For example, sync fails fwith such a schema for destination Bigquery
You can see the results with my proposed schema [here](https://github.com/airbytehq/airbyte/pull/21695#issuecomment-1414208540).

Reply:
was that test using `oneof: [{type: object}, ...]` or `type: [object, ...]`?

my main concern is that right now the `type: object` is explicitly false - e.g. `[1, 2, 3]` is a valid jsonb, but it won't match the json schema. Platform is planning to fail syncs with incorrect schemas, so this is putting us in a slightly awkward position

Reply:
> was that test using `oneof: [{type: object}, ...]` or `type: [object, ...]`?
> 
> my main concern is that right now the `type: object` is explicitly false - e.g. `[1, 2, 3]` is a valid jsonb, but it won't match the json schema. Platform is planning to fail syncs with incorrect schemas, so this is putting us in a slightly awkward position

i see. In case of platform will validate every record we definitely can not use current json schema
@edgao i have updated my PR with the following json schema
```
{
  "type": ["object", "array", "string", "number", "boolean"],
  "airbyte_type": "jsonb"
}
```
**Original Postgres table**

<img width="625" alt="postgres-original-json-table" src="https://user-images.githubusercontent.com/39538064/220162393-b79e3386-4c2d-43bb-b1ad-acfc1c11080a.png">

## Tests results

**Postgres Destination** ✅ 
JSONB ----> JSONB 

**MySQL Destination** ✅ 
JSONB ----> JSON 

**MSSQL Destination** ❌ 
JSONB ----> VARCHAR 

**_Integers and booleans values are null_**

<img width="391" alt="mssql-jsonb-int-and-bool-null" src="https://user-images.githubusercontent.com/39538064/220162994-e2173865-eb60-4ef1-aa19-a662e2c46d30.png">

**Redshift Destination** 
JSONB ----> SUPER 
Standart inserts ✅ 
 S3 staging ✅ 

**Snowflake Destination**
JSONB ----> VARIANT 
Internal Staging ✅ 
AWS S3 Staging (CSV) ✅
GCS Staging (CSV) ✅
 Azure blob Staging (CSV) ✅

**S3 Destination**
 CSV ✅
JSONL ✅ 
AVRO ✅ 
Parquet ✅

**Oracle Destination** ❌ 
JSONB ----> VARCHAR 

**_All values are null except booleans and integers_**

<img width="246" alt="oracle-jsonb-all-null-without-int-and-bool" src="https://user-images.githubusercontent.com/39538064/220163607-98c276ce-c9a6-493f-b6e4-8212bc01d15a.png">

**Clickhouse Destination** ❌ 
JSONB ----> Array(String) 

**_All values are empty arrays except original arrays_**

<img width="307" alt="clickhouse-jsonb-all-values-empty-except-arrays" src="https://user-images.githubusercontent.com/39538064/220163939-3f9ad103-e85e-417b-84c9-f2fcd35de070.png">

**BigQuery  Destination**
JSONB ----> STRING
Standard inserts ❌
 GCS Staging ❌

**_All values are null except original arrays_**

<img width="348" alt="bigquery-jsonb-all-null-except-arrays" src="https://user-images.githubusercontent.com/39538064/220164337-705d2517-29bb-4d6c-af21-b65014e05f35.png">

**BigQuery Denormalized Destination**
JSONB ----> STRING

 Standard inserts  ❌❌❌ **failed sync**
`Error while reading data, error message: JSON parsing error in row starting at position 0: Array specified for non-repeated field: test_column.`

GCS Staging - ❌❌❌ **failed sync**
`Caused by: com.google.cloud.bigquery.BigQueryException: Field test_column has incompatible types. Configured schema: string; Avro file: record.`

Summing up the test results, the problems that are observed in MSSQL Destination, Oracle Destination, ClickHouse Destination and BigQuery Destination are definitely normalization issues since data populated correctly into the airbyte_raw table

Reply:
@evantahler you mentioned that platform will start failing syncs with schema mismatches “soon” - how soon is that? I.e. imminent doom, a couple weeks, a couple months?

from vitalli's research it looks like there's a few paths forward here:
1. Use `{type: object}` - this schema is wrong but has the best outcome in normalization today
     1. We'd need to prioritize adding a proper JSONB type to the protocol, with varying urgency based on when syncs will start failing. This has a dependency on the protocol v1 rollout, which is expected to extend into Q2.
2. Use `{type: [object, array, number, ....]}` - this is a valid schema, but has worse output
     1. Maybe we can figure out how to make this work better in normalization? But there's already a lot of stuff going on in that module right now, and I don't know if we have bandwidth

option 1 is a bit more appealing, but if the timeline is "next month" then that changes things

Reply:
Update on "soon" - we decided not to go that route, but instead, open up Sentry issues whenever we find records that don't match the schema.  That, in turn, will be funneled into on-call issues to be worked on.  TBD on the priority, but GA connectors will likely be higher priority than the rest.

This is a Q1B effort for the Connector Ops team - "Track & fix schema validation errors"

Reply:
neat! Sounds like worst-case we'll have some noise in OC, which... I feel OK with, since supporting jsonb is pretty valuable. We can just silence the postgres alert when it fires. Once we have a protocol-native jsonb type, I think this largely stops mattering (since the platform will internally upgrade all messages to v2, so it shouldn't see the hacky protocol v0 schema)

@VitaliiMaltsev I think the `{type: object, oneOf: [...]}` schema is actually fine, sorry for the flip flop here

Reply:
@edgao reverted to json schema `{type: object, oneOf: [...]}`